File: packages/controls/src/resolve.ts
import type { AssuranceProfile, ControlDefinition, ResolvedAssuranceProfile } from '@l9/assurance-contracts';

export class ControlResolutionError extends Error {
  constructor(message: string) { super(message); this.name='ControlResolutionError'; }
}

export function resolveProfile(profile: AssuranceProfile, definitions: readonly ControlDefinition[]): ResolvedAssuranceProfile {
  const byKey=new Map(definitions.map((control)=>[key(control.id,control.version),control]));
  const selected=profile.controls.map((reference)=>{
    const control=byKey.get(key(reference.id,reference.version));
    if(!control)throw new ControlResolutionError(`Missing control ${reference.id}@${reference.version}`);
    return control;
  });
  const ordered=orderControls(selected);
  return Object.freeze({profile,controls:Object.freeze(ordered)});
}

export function orderControls(controls: readonly ControlDefinition[]): readonly ControlDefinition[] {
  const byId=new Map<string,ControlDefinition>();
  for(const control of controls){if(byId.has(control.id))throw new ControlResolutionError(`Duplicate control ${control.id}`);byId.set(control.id,control);}
  const visiting=new Set<string>(); const visited=new Set<string>(); const output:ControlDefinition[]=[];
  function visit(control:ControlDefinition):void{
    if(visited.has(control.id))return;
    if(visiting.has(control.id))throw new ControlResolutionError(`Control dependency cycle at ${control.id}`);
    visiting.add(control.id);
    for(const dependency of [...(control.dependencies??[])].sort((a,b)=>a.id.localeCompare(b.id))){
      const target=byId.get(dependency.id); if(!target)throw new ControlResolutionError(`Control ${control.id} depends on missing ${dependency.id}`);
      if(target.version!==dependency.version)throw new ControlResolutionError(`Control ${control.id} dependency version mismatch for ${dependency.id}`);
      visit(target);
    }
    visiting.delete(control.id); visited.add(control.id); output.push(control);
  }
  for(const control of [...controls].sort((a,b)=>a.id.localeCompare(b.id)))visit(control);
  return Object.freeze(output);
}
function key(id:string,version:string):string{return`${id}@${version}`;}

File: src/l9_assurance/controls/resolve.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from copy import deepcopy
from typing import Any


class ControlResolutionError(ValueError):
    """Raised when a profile cannot resolve to a deterministic control graph."""


def resolve_profile(
    profile: Mapping[str, Any], definitions: Sequence[Mapping[str, Any]]
) -> dict[str, Any]:
    by_key = {(item["id"], item["version"]): item for item in definitions}
    selected: list[Mapping[str, Any]] = []
    for reference in profile["controls"]:
        control = by_key.get((reference["id"], reference["version"]))
        if control is None:
            raise ControlResolutionError(
                f"Missing control {reference['id']}@{reference['version']}"
            )
        selected.append(control)
    return {"profile": deepcopy(dict(profile)), "controls": order_controls(selected)}


def order_controls(controls: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    by_id: dict[str, Mapping[str, Any]] = {}
    for control in controls:
        control_id = str(control["id"])
        if control_id in by_id:
            raise ControlResolutionError(f"Duplicate control {control_id}")
        by_id[control_id] = control
    visiting: set[str] = set()
    visited: set[str] = set()
    output: list[dict[str, Any]] = []

    def visit(control: Mapping[str, Any]) -> None:
        control_id = str(control["id"])
        if control_id in visited:
            return
        if control_id in visiting:
            raise ControlResolutionError(f"Control dependency cycle at {control_id}")
        visiting.add(control_id)
        for dependency in sorted(control.get("dependencies", []), key=lambda item: item["id"]):
            target = by_id.get(dependency["id"])
            if target is None:
                raise ControlResolutionError(
                    f"Control {control_id} depends on missing {dependency['id']}"
                )
            if target["version"] != dependency["version"]:
                raise ControlResolutionError(
                    f"Control {control_id} dependency version mismatch for {dependency['id']}"
                )
            visit(target)
        visiting.remove(control_id)
        visited.add(control_id)
        output.append(deepcopy(dict(control)))

    for control in sorted(controls, key=lambda item: item["id"]):
        visit(control)
    return output
